src/composables/usePerf.ts

```diff
@@ -0,0 +1,141 @@
+import { useTresContext } from '@tresjs/core'
+import type { Scene, WebGLRenderer } from 'three'
+import { calculateMemoryUsage } from '../utils'
+
+interface MemoryUsageData {
+  currentMem: number
+  averageMem: number
+  allocatedMem: number
+  maxMemory: number
+  accumulator: number[]
+  lastLoggedTime: number
+  logInterval: number
+}
+
+interface FpsData {
+  value: number
+  accumulator: number[]
+  lastLoggedTime: number
+  logInterval: number
+}
+
+// Define the structure of the performance data
+interface PerfData {
+  fps: FpsData
+  memory: MemoryUsageData // Adjust 'any' to a more specific type as needed
+  gl: WebGLRenderer | undefined
+  scene: Scene | undefined
+}
+
+// Define the structure of the store
+interface PerfStore {
+  [uuid: string]: { perf: PerfData }
+}
+
+export const PERF_CONTEXT_KEY = Symbol('PERF_CONTEXT_KEY')
+const DEFAULT_UUID = 'default'
+
+// Internal state
+const perfStore = reactive<PerfStore>({
+  [DEFAULT_UUID]: {
+    perf: {
+      fps: reactive({
+        value: 0,
+        accumulator: [],
+        lastLoggedTime: Date.now(),
+        logInterval: 1000,
+      }),
+      memory: reactive({
+        currentMem: 0,
+        averageMem: 0,
+        maxMemory: 0,
+        allocatedMem: 0,
+        accumulator: [],
+        lastLoggedTime: Date.now(),
+        logInterval: 1000,
+      }), // Placeholder for memory object
+      gl: undefined,
+      scene: undefined,
+    },
+  },
+})
+
+export function usePerfProvider(uuid: string = DEFAULT_UUID) {
+  provide(PERF_CONTEXT_KEY, perfStore)
+  return perfStore[uuid].perf
+}
+
+export interface TresPerfOptions {
+  renderer?: Ref<WebGLRenderer>
+}
+
+export function usePerf(options?: TresPerfOptions) {
+  const { renderer: TresRenderer, scene } = useTresContext()
+  const gl = options?.renderer || TresRenderer.value
+  if (!gl) {
+    throw new Error('Renderer not found')
+  }
+
+  const perf = perfStore[DEFAULT_UUID].perf
+  perf.gl = gl as WebGLRenderer
+  perf.scene = scene.value as Scene
+
+  const updateInterval = 100 // Update interval in milliseconds
+  const fps = useFps({ every: updateInterval }) 
+  const { isSupported, memory } = useMemory({ interval: updateInterval })
+
+  const maxFrames = 80
+
+  let lastUpdateTime = performance.now()
+
+  const updatePerformanceData = ({ timestamp }: { timestamp: number }) => {
+
+    // Update WebGL Memory Usage (Placeholder for actual logic)
+    // perf.memory.value = calculateMemoryUsage(gl)
+    perf.memory.allocatedMem = calculateMemoryUsage(unref(scene))
+  
+    // Update memory usage
+    if (timestamp - lastUpdateTime >= updateInterval) {
+      lastUpdateTime = timestamp
+
+      // Update FPS
+      perf.fps.accumulator.push(fps.value as never)
+
+      if (perf.fps.accumulator.length > maxFrames) {
+        perf.fps.accumulator.shift()
+      }
+
+      perf.fps.value = perf.fps.accumulator.reduce((a, b) => a + b, 0) / perf.fps.accumulator.length
```

Comment:
When looking at the FPS graph in the demo, I was confused as I was expecting the current(ish) framerate. But I see here that the displayed framerate is the average of all the samples – currently covering the last 8 seconds.

Personally, I'd prefer to shorten that. E.g., [stats-gl shows the previous second's average framerate.](https://github.com/RenaudRohlinger/stats-gl/blob/main/lib/main.ts#L323)